library/check_crypto_config.h

```diff
@@ -93,4 +93,16 @@
 #error "MBEDTLS_SSL_PROTO_TLS1_2 defined, but not all prerequisites"
 #endif
 
+/* TLS 1.3 requires separate HKDF parts from PSA */
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3) && \
+        !( defined(MBEDTLS_USE_PSA_CRYPTO) && defined(PSA_WANT_ALG_HKDF_EXTRACT) && defined(PSA_WANT_ALG_HKDF_EXPAND) )
```

Comment:
This makes it so that `MBEDTLS_USE_PSA_CRYPTO` is mandatory, and previously it was `MBEDTLS_PSA_CRYPTO_C`. If that's the case then the check in `check_config.h` has an unnecessary branch, based on
```
( !defined(MBEDTLS_USE_PSA_CRYPTO) && \
      !(
	( defined(MBEDTLS_MD_C) && 
		( defined(MBEDTLS_SHA256_C) || defined(MBEDTLS_SHA384_C) )
	) 
)
```

Reply:
Indeed it should be MBEDTLS_PSA_CRYPTO_C.

Reply:
I force-pushed the last commit. Please check if this is what you expected as I'm not sure if I correctly understood the second part.

Reply:
Ah, good catch @AndrzejKurek. I think expression everything using just "and", "or", "not" and parentheses is becoming hard to read, so I'm wondering if nested `#if`s would make things more readable (actually, I've been thinking about this for the last few iterations of this discussion):
```
#if defined(MBEDTLS_SSL_PROTO_TLS1_3) && !defined(MBEDTLS_USE_PSA_CRYPTO)
/* When USE_PSA_CRYPTO is not defined, we also need SHA-256 or SHA-384 via the legacy interface, including via the MD layer (in addition to having them in PSA, for HDKF among other things), as parts of the code (running handshake hash) are shared with TLS 1.2 */
#if !( defined(MBEDTLS_MD_C) && ( defined(MBEDTLS_SHA256_C) || defined(MBEDTLS_SHA384_C) ) )
#error "..."
#endif
#endif
```
Wdyt?

Reply:
@mpg - That's a great idea :)

Reply:
I understand that additional condition when `PSA_CRYPTO` is available is also needed:
```
#if defined(MBEDTLS_SSL_PROTO_TLS1_3) && defined(MBEDTLS_USE_PSA_CRYPTO)
#if !( defined(PSA_WANT_ALG_SHA_256) || defined(PSA_WANT_ALG_SHA_348)
#error "..."
#endif
#endif
```

Reply:
Nope, we always need `PSA_WANT_ALG_SHA_256` (or 384). Unconditionally, not just when `USE_PSA_CRYPTO` is enabled. That's because they're used for example in HKDF, which is always done using PSA, because it's pure TLS 1.3.

The only parts of TLS 1.3 that care about `USE_PSA_CRYPTO` are those that are shared with TLS 1.2. The parts that are 1.3-only don't care about this macro and just always use PSA.

See my suggestion [here](https://github.com/Mbed-TLS/mbedtls/pull/6255#discussion_r968217668) (which was previously hidden as part of a resolved discussion, which I just "un-resolved" to make it more visible.